# Waze User Churn Exploratory Data Analysis

## Objective
The goal of this analysis is to explore user driving behavior data from Waze and identify patterns associated with user churn.

## Key Questions
1. Are there missing values and do they follow a pattern?
2. How do churned users differ from retained users?
3. Does device type influence churn?
4. What behavioral metrics best explain churn?

### **01. Imports and data loading**

In [ ]:
# Import packages for data manipulation
import pandas as pd
import numpy as np


In [ ]:
# Load dataset into dataframe
df = pd.read_csv('waze_dataset.csv')

FileNotFoundError: [Errno 2] No such file or directory: 'waze_dataset.csv'

### **02. Summary information**

In [ ]:
df.head(10)


,ID,label,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days,device
0,0,retained,283,226,296.748273,2276,208,0,2628.845068,1985.775061,28,19,Android
1,1,retained,133,107,326.896596,1225,19,64,13715.920550,3160.472914,13,11,iPhone
2,2,retained,114,95,135.522926,2651,0,0,3059.148818,1610.735904,14,8,Android
3,3,retained,49,40,67.589221,15,322,7,913.591123,587.196542,7,3,iPhone
4,4,retained,84,68,168.247020,1562,166,5,3950.202008,1219.555924,27,18,Android
5,5,retained,113,103,279.544437,2637,0,0,901.238699,439.101397,15,11,iPhone
6,6,retained,3,2,236.725314,360,185,18,5249.172828,726.577205,28,23,iPhone
7,7,retained,39,35,176.072845,2999,0,0,7892.052468,2466.981741,22,20,iPhone
8,8,retained,57,46,183.532018,424,0,26,2651.709764,1594.342984,25,20,Android
9,9,churned,84,68,244.802115,2997,72,0,6043.460295,2341.838528,7,3,iPhone


In [ ]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


## Dataset
The dataset contains **14,999 users** and **13 behavioral variables**, including:
- sessions
- drives
- kilometers driven
- driving duration
- activity days
- device type

### **03. Null values and summary statistics**


In [ ]:
# rows with null labels
null_rows = df[df['label'].isna()]

# summary stats
null_rows.describe()

,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
count,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000
mean,7405.584286,80.837143,67.798571,198.483348,1709.295714,118.717143,30.371429,3935.967029,1795.123358,15.382857,12.125714
std,4306.900234,79.987440,65.271926,140.561715,1005.306562,156.308140,46.306984,2443.107121,1419.242246,8.772714,7.626373
min,77.000000,0.000000,0.000000,5.582648,16.000000,0.000000,0.000000,290.119811,66.588493,0.000000,0.000000
25%,3744.500000,23.000000,20.000000,94.056340,869.000000,4.000000,0.000000,2119.344818,779.009271,8.000000,6.000000
50%,7443.000000,56.000000,47.500000,177.255925,1650.500000,62.500000,10.000000,3421.156721,1414.966279,15.000000,12.000000
75%,11007.000000,112.250000,94.000000,266.058022,2508.750000,169.250000,43.000000,5166.097373,2443.955404,23.000000,18.000000
max,14993.000000,556.000000,445.000000,1076.879741,3498.000000,1096.000000,352.000000,15135.391280,9746.253023,31.000000,30.000000


In [ ]:
# rows without null labels
non_null_rows = df[df['label'].notna()]

# summary stats
non_null_rows.describe()

,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
count,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000
mean,7503.573117,80.623820,67.255822,189.547409,1751.822505,121.747395,29.638296,4044.401535,1864.199794,15.544653,12.182530
std,4331.207621,80.736502,65.947295,136.189764,1008.663834,147.713428,45.350890,2504.977970,1448.005047,9.016088,7.833835
min,0.000000,0.000000,0.000000,0.220211,4.000000,0.000000,0.000000,60.441250,18.282082,0.000000,0.000000
25%,3749.500000,23.000000,20.000000,90.457733,878.500000,10.000000,0.000000,2217.319909,840.181344,8.000000,5.000000
50%,7504.000000,56.000000,48.000000,158.718571,1749.000000,71.000000,9.000000,3496.545617,1479.394387,16.000000,12.000000
75%,11257.500000,111.000000,93.000000,253.540450,2627.500000,178.000000,43.000000,5299.972162,2466.928876,23.000000,19.000000
max,14998.000000,743.000000,596.000000,1216.154633,3500.000000,1236.000000,415.000000,21183.401890,15851.727160,31.000000,30.000000


The summary statistics of the rows with missing labels are very similar to those of the rows without missing values. Metrics such as sessions, drives, kilometers driven, and driving duration show comparable averages between the two groups. This suggests that the missing label values are likely random and not associated with a specific pattern in the dataset.

| Metric                  | Null label mean | Non-null mean |
| ----------------------- | --------------- | ------------- |
| sessions                | ~80.84          | ~80.62        |
| drives                  | ~67.80          | ~67.26        |
| driven_km_drives        | ~3936 km        | ~4044 km      |
| duration_minutes_drives | ~1795           | ~1864         |

### **04. Null values - device counts**

In [ ]:
# Get count of null values by device
null_rows['device'].value_counts()

,count
device,
iPhone,447
Android,253


Among the rows with missing values in the label column, 447 users were using an iPhone and 253 users were using Android devices.

In [ ]:
# Calculate % of iPhone nulls and Android nulls
null_rows['device'].value_counts(normalize=True)


,proportion
device,
iPhone,0.638571
Android,0.361429


In [ ]:
# Calculate % of iPhone users and Android users in full dataset
df['device'].value_counts(normalize=True)


,proportion
device,
iPhone,0.644843
Android,0.355157


The percentage of missing values by each device is consistent with their representation in the data overall.

There is nothing to suggest a non-random cause of the missing data.

Counts and percentages of users who churned vs. those who were retained.

In [ ]:
# Calculate counts of churned vs. retained
df['label'].value_counts()



,count
label,
retained,11763
churned,2536


In [ ]:
df['label'].value_counts(normalize=True)

,proportion
label,
retained,0.822645
churned,0.177355


This dataset contains 82% retained users and 18% churned users.

"Median" comparison of each variable for churned and retained users.

In [ ]:
# Calculate median values of all columns for churned and retained users
df.groupby('label').median(numeric_only=True)

,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
label,,,,,,,,,,,
churned,7477.5,59.0,50.0,164.339042,1321.0,84.5,11.0,3652.655666,1607.183785,8.0,6.0
retained,7509.0,56.0,47.0,157.586756,1843.0,68.0,9.0,3464.684614,1458.046141,17.0,14.0


This offers an interesting snapshot of the two groups, churned vs. retained:

Users who churned averaged ~3 more drives in the last month than retained users, but retained users used the app on over twice as many days as churned users in the same time period.

The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

It seems that churned users had more drives in fewer days, and their trips were farther and longer in duration. Perhaps this is suggestive of a user profile.

"Median" kilometers per drive in the last month for both retained and churned users.

In [ ]:
# Add a column to df called `km_per_drive`
df['km_per_drive'] = df['driven_km_drives'] / df['drives']

# Group by `label`, calculate the median, and isolate for km per drive
df.groupby('label')['km_per_drive'].median()


,km_per_drive
label,
churned,74.109416
retained,75.014702


The median retained user drove about one more kilometer per drive than the median churned user.

How many kilometers per driving day was this?

In [ ]:
# Add a column to df called `km_per_driving_day`
df['km_per_driving_day'] = df['driven_km_drives'] / df['driving_days']

# Group by `label`, calculate the median, and isolate for km per driving day
df.groupby('label')['km_per_driving_day'].median()

,km_per_driving_day
label,
churned,697.541999
retained,289.549333


"Median" number of drives per driving day for each group.

In [ ]:
# Add a column to df called `drives_per_driving_day`
df['drives_per_driving_day'] = df['drives'] / df['driving_days']

# Group by `label`, calculate the median, and isolate for drives per driving day
df.groupby('label')['drives_per_driving_day'].median()

,drives_per_driving_day
label,
churned,10.0000
retained,4.0625


The median user who churned drove 698 kilometers each day they drove last month, which is almost ~240% the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers! It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data&mdash;and in particular the sample of churned users&mdash;contains a high proportion of long-haul truckers.

In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

Check for imbalance in how many users churned by device type?

In [ ]:
# For each label, calculate the number of Android users and iPhone users
df.groupby(['label', 'device']).size()

label     device 
churned   Android     891
          iPhone     1645
retained  Android    4183
          iPhone     7580
dtype: int64

In [ ]:
# For each label, calculate the percentage of Android users and iPhone users
df.groupby('label')['device'].value_counts(normalize=True)


label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: proportion, dtype: float64

The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.

The distribution of device types is very similar between churned and retained users. Approximately 65% of users in both groups use iPhones and about 35% use Android devices. Because these proportions are nearly identical, device type does not appear to significantly influence whether a user churns or remains active.

### **Key Findings**

1. The dataset contains missing values in the label column. Out of 14,999 rows, 700 rows were missing churn labels. When comparing users with missing labels to those with complete data, the summary statistics were very similar. Additionally, the distribution of device types among rows with missing values closely matched the distribution in the overall dataset. This suggests that the missing data is likely random rather than systematic.

2. The median is less sensitive to extreme values and outliers. In this dataset, some users drove extremely large distances, which could distort the mean. Using the median provides a more accurate representation of the typical user behavior.

3. The analysis suggests that many users in the dataset may be heavy or long-distance drivers, possibly professional drivers such as truck or delivery drivers. It would be useful to investigate whether these users have different navigation needs compared to typical commuters. Additional information about user occupation, trip purpose, or driving context could help better understand why some users churn.

4. Approximately 35% of users were Android users and about 65% were iPhone users in the dataset.

5. Churned users tended to:

    Drive longer distances per day

    Have more drives per driving day

    Spend more total time driving

   However, retained users used the app on more days overall, suggesting more consistent engagement with the application.

6. The proportion of Android and iPhone users was nearly identical between churned and retained users. This indicates that device type does not appear to significantly influence churn.